Log Workspaces supported sql functions (#26042)
DataDog/documentation: content/en/logs/workspaces/sql_reference.md

---
title: SQL Reference
further_reading:
- link: "/logs/workspaces/"
  tag: "Documentation"
  text: "Learn more about Log Workspaces"
---

## Overview

SQL in Analysis cells allows you to analyze and manipulate data within Log Workspaces. This documentation covers the SQL support available in Log Workspaces and includes:
- [SQL functions](#functions)
- [Syntax compatible with PostgreSQL](#syntax)

{{< img src="/logs/workspace/sql_reference/sql_syntax_analysis_cell.png" alt="Example workspace cell with SQL syntax" style="width:100%;" >}}

## Functions

The following SQL functions are supported:

> | Function         | Description                                                        |

Reviewer: For a function reference like this, I would suggest having more information about the actual signature of the functions (and/or specifying that these are just the same as the equivalent Postgres functions).

As an example of this (which could also be aligned/reused here) is the [existing DDSQL Editor docs](https://docs.datadoghq.com/ddsql_editor/reference/functions/), which provide the full signature, as well as having the split between scalar and aggregation functions (which is also very relevant).

Author: Thanks @gabrielreid, would you be able to provide an example, such as for the `MIN` function, what columns are you looking for and values.

Author: @gabrielreid I added examples for Syntax and Functions, is this close to what you're thinking?

|------------------|--------------------------------------------------------------------|
| `MIN`            | Returns the minimum value across all input values                  |
| `MAX`            | Returns the maximum value across all input values                  |
| `COUNT`          | Returns the number of input values that are not null               |
| `SUM`            | Returns the summation across all input values                      |
| `AVG`            | Returns the average value (arithmetic mean) across all input values|
| `CEIL / CEILING` | Returns the value rounded up to the nearest integer                |
| `FLOOR`          | Returns the value rounded down to the nearest integer              |
| `ROUND`          | Returns the value rounded to the nearest integer                   |
| `LOWER`          | Returns the string as lower case                                   |
| `UPPER`          | Returns the string as uppercase                                    |
| `ABS`            | Returns the absolute value                                         |
| `COALESCE`       | Returns the first non-null value                                   |

> ## Syntax

Reviewer: Note that the heading above that links to this specifies "Postgres-compatible syntax", but this does not. I think it's worth making it clear that this syntax is specifically in line with either Postgres, or "standard" SQL.

Here also, it's worth having more concrete examples of these various operators such as `LIKE` and `CASE`. I would really suggest having a few working, concrete examples here.

Author: ```suggestion
## Syntax

The following syntax is compatible with PostgreSQL.

```

* `SELECT (DISTINCT)`  
* `JOIN`  
* `GROUP BY`  
* `WHERE` (including support for `LIKE` filters on strings)  
* `CASE`   
* Basic (standard) arithmetic operations (such as `+`,`-`, `*`, `/`)

## Casting

To cast values in columns, select between types when defining them in a Logs data source:

> 1. In a workspace cell, click the cog labeled **Columns**.

Reviewer: Note that this is only for a logs data source cell, not for _any_ workspace cell.

Author: ```suggestion
1. In a Logs Data Source cell, click the cog labeled **Columns**.
```

---
title: SQL Reference
further_reading:
- link: "/logs/workspaces/"
  tag: "Documentation"
  text: "Learn more about Log Workspaces"
---

## Overview

> SQL functions are built-in operations in SQL (Structured Query Language) used for tasks like calculations, data manipulation, or formatting. This documentation is a reference for SQL support available in Logs Workspaces. It details:

Reviewer 2: Nit: I'm not sure whether we should start this guide with a note about explaining SQL functions since this is a reference guide on SQL support as a whole.

- [SQL functions](#functions)
- [Syntax compatible with PostgreSQL](#syntax)

> It provides a resource for the Analysis cell. Use this information to leverage SQL capabilities in data analysis and manipulation within Log Workspaces.

Reviewer 2: While reading the doc, I found the following line a bit out of context.
> It provides a resource for the Analysis cell.

Should we not be leading with this? That the documentation provides a reference for the capabilities of the SQL Analysis cell?

| `MAX`            | Returns the maximum value across all input values                  |
| `COUNT`          | Returns the number of input values that are not null               |
| `SUM`            | Returns the summation across all input values                      |
| `AVG`            | Returns the average value (arithmetic mean) across all input values|
| `CEIL / CEILING` | Returns the value rounded up to the nearest integer                |
| `FLOOR`          | Returns the value rounded down to the nearest integer              |
| `ROUND`          | Returns the value rounded to the nearest integer                   |
| `LOWER`          | Returns the string as lower case                                   |
| `UPPER`          | Returns the string as uppercase                                    |
| `ABS`            | Returns the absolute value                                         |
| `COALESCE`       | Returns the first non-null value                                   |

## Syntax

* `SELECT (DISTINCT)`  
* `JOIN`  
* `GROUP BY`  
* `WHERE` (including support for `LIKE` filters on strings)  
* `CASE`   

> * Basic (standard) arithmetic operations (e.g. \+, \-, \*, /)

Author: ```suggestion
* Basic (standard) arithmetic operations (such as `+`,`-`, `*`, `/`)
```

## Casting

> To cast values in columns, select between types when defining them in a Logs data source:

Reviewer 2: I'm not sure whether this front-end type casting of a logs datasource should be in this SQL reference guide. Should this rather be an explanation of the _SQL_ cast that's soon to be available with a description of the supported types?

Reviewer: Definitely agree with what Vishal is saying here. If this page is about SQL, this section probably doesn't belong here (and even if it does belong here, I would propose using an alternate heading like "Data types" instead of casting, because what's described here is not really casting).

Author: Will remove the casting section!

- [SQL functions](#functions)
- [Data types](#data-types)
- [Syntax compatible with PostgreSQL](#syntax)

It provides a resource for the Analysis cell. Use this information to leverage SQL capabilities in data analysis and manipulation within Log Workspaces.

{{< img src="/logs/workspace/sql_reference/sql_syntax_analysis_cell.png" alt="Example workspace cell with SQL syntax" style="width:100%;" >}}

## Functions

The following SQL functions are supported:

* `MIN` (returns the minimum value across all input values)  
* `MAX` (returns the maximum value across all input values)  
* `COUNT` (returns the number of input values that are not null)  
* `SUM` (returns the summation across all input values)  
* `AVG` (returns the average value (arithmetic mean)  across all input values)  
    
* `CEIL / CEILING` (returns the value rounded up to the nearest integer)  
* `FLOOR` (returns the value rounded down to the nearest integer)  
* `ROUND` (returns the value rounded to the nearest integer)  
* `LOWER` (returns the string as lower case)  
* `UPPER` (returns the string as uppercase)  
* `ABS` (returns the absolute value)  
    

> * `Coalesce` (returns the first non-null value)

Author: ```suggestion
* `COALESCE` (returns the first non-null value)
```

FROM logs 
WHERE status_code = 200 {{< /code-block >}} |
| `MAX(numerical column)`       | Returns the maximum value across all input values.                   | {{< code-block lang="sql" >}}SELECT MAX(response_time) AS max_response_time 
FROM logs 
WHERE status_code = 200 {{< /code-block >}} |
| `COUNT(any column)`     | Returns the number of input values that are not null.                      | {{< code-block lang="sql" >}}SELECT COUNT(request_id) AS total_requests 
FROM logs 
WHERE status_code = 200 {{< /code-block >}} |
| `SUM(numerical column)`       | Returns the summation across all input values.                       | {{< code-block lang="sql" >}}SELECT SUM(bytes_transferred) AS total_bytes 
FROM logs 
GROUP BY service_name {{< /code-block >}} |
| `AVG(numerical column)`       | Returns the average value (arithmetic mean) across all input values. | {{< code-block lang="sql" >}}SELECT AVG(response_time) 
AS avg_response_time 
FROM logs 
WHERE status_code = 200 
GROUP BY service_name {{< /code-block >}} |

> | `CEIL`/`CEILING(numerical column)` | Returns the value rounded up to the nearest integer.            | {{< code-block lang="sql" >}} ELECT CEIL(price) AS rounded_price 

Author: ```suggestion
| `CEIL`/`CEILING(numerical column)` | Returns the value rounded up to the nearest integer.            | {{< code-block lang="sql" >}} SELECT CEIL(price) AS rounded_price 
```
